lib/lambda_ethereum_consensus/ssz_ex/encode.ex

```diff
@@ -0,0 +1,187 @@
+
+      with {:ok, {encoded_variable_parts, variable_offsets_list, total_byte_size}} <-
+             encode_variable_parts(list, inner_type),
+           :ok <- check_length(fixed_lengths, total_byte_size),
+           {variable_offsets, _} =
+             Enum.reduce(variable_offsets_list, {[], 0}, fn element, {res, acc} ->
+               sum = fixed_lengths + acc
+               {[sum | res], element + acc}
+             end),
+           {:ok, encoded_variable_offsets} <-
+             variable_offsets
+             |> Enum.reverse()
+             |> Enum.map(&encode(&1, {:int, 32}))
+             |> Utils.flatten_results() do
+        (encoded_variable_offsets ++ encoded_variable_parts)
+        |> :binary.list_to_bin()
+        |> then(&{:ok, &1})
+      end
+    end
+  end
+
+  defp encode_variable_parts(list, inner_type) do
+    with {:ok, {encoded_list, byte_size_list, total_byte_size}} <-
+           Enum.reduce_while(list, {:ok, {[], [], 0}}, fn value,
+                                                          {:ok, {res_encoded, res_size, acc}} ->
+             case encode(value, inner_type) do
+               {:ok, encoded} ->
+                 size = byte_size(encoded)
+                 {:cont, {:ok, {[encoded | res_encoded], [size | res_size], size + acc}}}
+
+               error ->
+                 {:halt, {:error, error}}
+             end
+           end) do
+      {:ok, {Enum.reverse(encoded_list), Enum.reverse(byte_size_list), total_byte_size}}
+    end
+  end
+
+  defp encode_container(container, schemas) do
+    {fixed_size_values, fixed_length, variable_values} = analyze_schemas(container, schemas)
+
+    with {:ok, variable_parts} <- encode_schemas(Enum.reverse(variable_values)),
+         offsets = calculate_offsets(variable_parts, fixed_length),
+         variable_length =
+           Enum.reduce(variable_parts, 0, fn part, acc -> byte_size(part) + acc end),
+         :ok <- check_length(fixed_length, variable_length),
+         {:ok, fixed_parts} <-
+           replace_offsets(fixed_size_values, offsets)
+           |> encode_schemas() do
+      (fixed_parts ++ variable_parts)
+      |> Enum.join()
+      |> then(&{:ok, &1})
+    end
+  end
+
+  defp analyze_schemas(container, schemas) do
+    schemas
+    |> Enum.reduce({[], 0, []}, fn {key, schema},
+                                   {acc_fixed_size_values, acc_fixed_length, acc_variable_values} ->
+      value = Map.fetch!(container, key)
+
+      if Utils.variable_size?(schema) do
+        {[:offset | acc_fixed_size_values], @bytes_per_length_offset + acc_fixed_length,
+         [{value, schema} | acc_variable_values]}
+      else
+        {[{value, schema} | acc_fixed_size_values],
+         acc_fixed_length + Utils.get_fixed_size(schema), acc_variable_values}
+      end
+    end)
+  end
+
+  defp encode_schemas(tuple_values) do
+    Enum.map(tuple_values, fn {value, schema} -> encode(value, schema) end)
```

Comment:
It would be nice that when we encode/decode or hash schemas and one field fails, then the field is mentioned in the error message.